snuba/migrations/table_engines.py

```diff
@@ -12,11 +14,21 @@ class TableEngine(ABC):
     """
 
     @abstractmethod
-    def get_sql(self) -> str:
+    def get_sql(self, context: OperationContext) -> str:
         raise NotImplementedError
 
 
 class MergeTree(TableEngine):
+    """
+    Provides the SQL for create table operations.
+
+    If the operation is being performed on a cluster marked as multi node, the
+    replicated versions of the tables will be created instead of the regular ones.
+
+    The layer, shard and replica values will be taken from the macros configuration
+    for multi node clusters, so these need to be set prior to running the migration.
```

Comment:
So let me see if I get the idea correctly.
If your cluster is multi node at the time of creating a table we go for a replicated one otherwise we go for a standalone one. Is this correct ?

A few questions:
- will there be a separate migration to create the distributed table on the query nodes?
- how does the migrations table work here? Is it going to be distributed ?
- do you plan to add some checks to ensure that the sinngle_node flag does not change later on (imagine when you are creating a new table)?
- All migrations are ran individually on every node, is that correct ?

Reply:
> * will there be a separate migration to create the distributed table on the query nodes?

In addition to the current forwards_local and backwards_local methods, each migration should also get the forwards_dist and backwards_dist methods, which will specify the set of operations to be run on the distributed nodes. It needs to be part of the same migration not a separate one as the local operations as these additional operations are necessary to  bring the system to the equivalent functioning state and should not be skipped or run at a different time.

> * how does the migrations table work here? Is it going to be distributed ?

For Sentry, most likely just replicated but not distributed. But ultimately, this is configuration provided by the user and our codebase can support both. The migration itself will be updated to provide all of the relevant distributed commands as well. So the table can be distributed or just replicated or not replicated at all depending on the users configuration.

> * do you plan to add some checks to ensure that the sinngle_node flag does not change later on (imagine when you are creating a new table)?

This is tricky to detect, since we are only really aware of the current value of that setting, not its prior state. I think we could potentially extend the schema diffing script to handle non single node clusters though. We could then ensure this passes before running any DDL operation.

> * All migrations are ran individually on every node, is that correct ?

Right now - yes. But certain commands don't actually need to be run on each replica. Specifically INSERT and ALTER queries are automatically replicated, but all the others are not. So instead of running them on all hosts we can potentially just run them on the first replica of each shard.

Reply:
re: distributed node. Ok it makes sense.
re: migrations table. Mind adding a TODO to update the migration so we keep track of the direction ?
re: you could verify that all tables on the database are either replicated or non replicated before running migrations and that is compatible with the current state of the cluster.
re: running invidivually on every node. Ok, makes sense.